tensorflow/core/platform/posix/posix_file_system.cc

```diff
@@ -276,4 +281,57 @@ Status PosixFileSystem::RenameFile(const string& src, const string& target) {
   return result;
 }
 
+Status PosixFileSystem::CopyFile(const string& src, const string& target) {
+  string translated_src = TranslateName(src);
+  struct stat sbuf;
+  if (stat(translated_src.c_str(), &sbuf) != 0) {
+    return IOError(src, errno);
+  }
+  int src_fd = open(translated_src.c_str(), O_RDONLY);
+  if (src_fd < 0) {
+    return IOError(src, errno);
+  }
+  string translated_target = TranslateName(target);
+  // O_WRONLY | O_CREAT:
+  //   Open file for write and if file does not exist, create the file.
+  // S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH:
+  //   Create the file with permission of 0666
+  int target_fd =
+      open(translated_target.c_str(), O_WRONLY | O_CREAT,
+           S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
+  if (target_fd < 0) {
+    close(src_fd);
+    return IOError(target, errno);
+  }
+  int rc = 0;
+  off_t offset = 0;
+  while (offset < sbuf.st_size) {
+    size_t chunk = (sbuf.st_size - offset) < SSIZE_MAX ? (sbuf.st_size - offset)
```

Comment:
Mixing signed and unsigned with integer limits like this scares me a little bit because I can't tell if the code is correct. Would it be possible to write this in a way there the code is obviously correct? Even if it results in a ton of static_cast<> boilerplate? I'll also note that I would personally feel completely fine if sendfile() context-switched back into userspace every 128kB. What matters IIRC is that the memory itself isn't being copied between the two.

Reply:
Done.